client/authority-discovery/src/worker.rs

```diff
@@ -247,6 +245,76 @@ where
 		}
 	}
 
+	/// Start the worker
+	pub async fn run(mut self) {
+		loop {
+			self.start_new_lookups();
+
+			futures::select! {
+				// Process incoming events.
+				event = self.dht_event_rx.next().fuse() => {
+					if let Some(event) = event {
+						self.handle_dht_event(event).await;
+					} else {
+						// This point is reached if the network has shut down, at which point there is not
+						// much else to do than to shut down the authority discovery as well.
+						return;
+					}
+				},
+				// Handle messages from [`Service`].
+				msg = self.from_service.next().fuse() => {
+					if let Some(msg) = msg {
+						self.process_message_from_service(msg);
+					} else {
+						return;
```

Comment:
An authority discovery `Service` allows other components to interact with the authority discovery worker. This features is not used as of today, but will be in the future within Polkadot. Given that it is not used today the sender side of the `from_service` channel will be closed at start-up. The worker should not return in case it is closed, but continue without instead.

Would you mind applying the following diff adding both a test as well as fixing the bug?

```diff
diff --git a/client/authority-discovery/src/worker.rs b/client/authority-discovery/src/worker.rs
index c18870003..44b674e46 100644
--- a/client/authority-discovery/src/worker.rs
+++ b/client/authority-discovery/src/worker.rs
@@ -261,13 +261,9 @@ where
                                                return;
                                        }
                                },
-                               // Handle messages from [`Service`].
-                               msg = self.from_service.next().fuse() => {
-                                       if let Some(msg) = msg {
-                                               self.process_message_from_service(msg);
-                                       } else {
-                                               return;
-                                       }
+                               // Handle messages from [`Service`]. Ignore if sender side is closed.
+                               msg = self.from_service.select_next_some() => {
+                                       self.process_message_from_service(msg);
                                },
                                // Set peerset priority group to a new random set of addresses.
                                _ = self.priority_group_set_interval.next().fuse() => {
diff --git a/client/authority-discovery/src/worker/tests.rs b/client/authority-discovery/src/worker/tests.rs
index b19914a9b..a7eccf4e3 100644
--- a/client/authority-discovery/src/worker/tests.rs
+++ b/client/authority-discovery/src/worker/tests.rs
@@ -441,8 +441,10 @@ fn publish_discover_cycle() {
        pool.run();
 }
 
+/// Don't terminate when sender side of service channel is dropped. Terminate when network event
+/// stream terminates.
 #[test]
-fn terminate_when_event_stream_terminates() {
+fn termination() {
        let (dht_event_tx, dht_event_rx) = channel(1000);
        let network: Arc<TestNetwork> = Arc::new(Default::default());
        let key_store = KeyStore::new();
@@ -450,7 +452,7 @@ fn terminate_when_event_stream_terminates() {
                authorities: vec![],
        });
 
-       let (_to_worker, from_service) = mpsc::channel(0);
+       let (to_worker, from_service) = mpsc::channel(0);
        let worker = Worker::new(
                from_service,
                test_api,
@@ -465,10 +467,17 @@ fn terminate_when_event_stream_terminates() {
        block_on(async {
                assert_eq!(Poll::Pending, futures::poll!(&mut worker));
 
+               // Drop sender side of service channel.
+               drop(to_worker);
+               assert_eq!(
+                       Poll::Pending, futures::poll!(&mut worker),
+                       "Expect the authority discovery module not to terminate once the \
+                        sender side of the service channel is closed.",
+               );
+
                // Simulate termination of the network through dropping the sender side
                // of the dht event channel.
                drop(dht_event_tx);
-
                assert_eq!(
                        Poll::Ready(()), futures::poll!(&mut worker),
                        "Expect the authority discovery module to terminate once the \
```

Reply:
Applied. Thanks!